### ISSUE TITLE
vmidbg fills up the host's memory until it's oom killed

### ISSUE THREAD

**Issue by waterbear515** [KVM-VMI/kvm-vmi#88]
Hello, thanks so much for your work on this project. It's very inspiring, and I hope to make good use of it after I find a way to resolve this issue.

### Issue:
I've followed the "Bare Metal" setup here (https://kvm-vmi.github.io/kvm-vmi/master/setup.html) exactly, and believe I have successfully installed all the necessary parts. I've verified this by running `hookguest-libkvmi /tmp/introspector`

```
grim@reverse:~/pyvmidbg$ sudo hookguest-libkvmi /tmp/introspector 
Waiting...
New handshake: name 'win10' start_time 1603591319
New guest: 3c 6c 05 06 47 93 4a 60 b9 a3 e5 98 84 4b 8b 73 fd 6 ctx (nil)
Sending the pause command...
We should receive 6 pause events
Max gfn: 0xfc816
Waiting...
Pop event
Starting #PF test with CR3 0x1aa002 (vcpu0)
Set page access gpa 0x1aa000 access r-x [0x5]
...
```

I can even get run `vmi-win-guid`, or dump the memory with `vmi-dump-memory`:
```
sudo ./examples/vmi-win-guid name win10 /tmp/introspector 
Windows Kernel found @ 0x2c00000
	Version: 64-bit Windows 10
	PE GUID: a371a2e91046000
	PDB GUID: 641f55c592201dcc4f59facc72ea54da1
	Kernel filename: ntkrnlmp.pdb
	Multi-processor without PAE
	Signature: 17744.
	Machine: 34404.
	# of sections: 33.
	# of symbols: 0.
	Timestamp: 2742133481.
	Characteristics: 34.
	Optional header size: 240.
	Optional header type: 0x20b
	Section 1: .rdata
	Section 2: .pdata
	Section 3: .idata
	Section 4: .edata
	Section 5: PROTDATA
	Section 6: GFIDS
	Section 7: Pad1
	Section 8: .text
	Section 9: PAGE
	Section 10: PAGELK
	Section 11: POOLCODE
	Section 12: PAGEKD
	Section 13: PAGEVRFY
	Section 14: PAGEHDLS
	Section 15: PAGEBGFX
	Section 16: INITKDBG
	Section 17: TRACESUP
	Section 18: KVASCODE
	Section 19: RETPOL
	Section 20: MINIEX
	Section 21: INIT
	Section 22: Pad2
	Section 23: .data
	Section 24: ALMOSTRO
	Section 25: CACHEALI
	Section 26: PAGEDATA
	Section 27: PAGEVRFD
	Section 28: INITDATA
	Section 29: Pad3
	Section 30: CFGRO
	Section 31: Pad4
	Section 32: .rsrc
	Section 33: .reloc
```

However, when I run `sudo python3 -m vmidbg -a 0.0.0.0 5000 win10 -k /tmp/introspector mspaint -d` and connect to it with `r2 gdb://127.0.0.1:5000`, it waits for about 30s until it rapidly starts consuming all the memory on the host and is killed by oom killer. Crazy, right? I've traced the issue all the way to `vmi_init_complete` within the libvmi library; it gets oom killed right after that function runs. (The computer has 12GB of memory and it's no way it's actually running out of it legitimately).

**Here's a short video of this happening:** https://i.imgur.com/ROYOhxs.mp4

**Debug Output:**
```
grim@reverse:~/pyvmidbg$ sudo python3 -m vmidbg -a 0.0.0.0 5000 win10 -k /tmp/introspector mspaint -d
INFO:server:listening on 0.0.0.0:5000
DEBUG:server:ready for next client
INFO:server:new client ('127.0.0.1', 36748)
--libkvmi path: /usr/local/lib/libkvmi.so
--Destroying KVM driver
--found KVM
--libkvmi path: /usr/local/lib/libkvmi.so
--KVMi socket path: /tmp/introspector
--libvirt version 6000000
--Connecting to KVMI...
--KVMi handshake:
--    VM name: win10
--    VM start time: 19:19:29 - Sat Oct 24 2020
--KVMi new guest:
--    UUID: 3C 6C 05 06 47 93 4A 60 B9 A3 E5 98 84 4B 8B 73
--    FD: 11
--    Protocol version: 1
--KVMi features:
--    VMFUNC: No
--    EPTP: No
--    VE: No
--    SPP: No
--KVMI connected
--VCPU count: 6
--Reading register 72 not implemented
Killed
```

@Wenzel has this ever happened to you? Do you have any idea what could be happening here?

Interestingly, this was happening on my AMD computer too, until I made the libvmi.conf use a rekall profile. However, on this Intel computer, no matter what I do I can't get it to work. I've tried every combination of rekall, volatility, and windows offsets imaginable, but it doesn't seem to fix it in the same way.

### Info:
**CPU:**
```
01: None 00.0: 10103 CPU                                        
  [Created at cpu.465]
  Unique ID: rdCR.j8NaKXDZtZ6
  Hardware Class: cpu
  Arch: X86-64
  Vendor: "GenuineIntel"
  Model: 6.62.4 "Intel(R) Core(TM) i7-4930K CPU @ 3.40GHz"
  Features: fpu,vme,de,pse,tsc,msr,pae,mce,cx8,apic,sep,mtrr,pge,mca,cmov,pat,pse36,clflush,dts,acpi,mmx,fxsr,sse,sse2,ss,ht,tm,pbe,syscall,nx,pdpe1gb,rdtscp,lm,constant_tsc,arch_perfmon,pebs,bts,rep_good,nopl,xtopology,nonstop_tsc,cpuid,aperfmperf,pni,pclmulqdq,dtes64,monitor,ds_cpl,vmx,est,tm2,ssse3,cx16,xtpr,pdcm,pcid,dca,sse4_1,sse4_2,x2apic,popcnt,tsc_deadline_timer,aes,xsave,avx,f16c,rdrand,lahf_lm,cpuid_fault,epb,pti,ssbd,ibrs,ibpb,stibp,tpr_shadow,vnmi,flexpriority,ept,vpid,fsgsbase,smep,erms,xsaveopt,dtherm,ida,arat,pln,pts,md_clear,flush_l1d
  Clock: 2007 MHz
  BogoMips: 6804.57
  Cache: 12288 kb
  Units/Processor: 32
  Config Status: cfg=new, avail=yes, need=no, active=unknown
```
**Kernel:**
```
Linux reverse 5.4.24+ #2 SMP Sat Oct 24 15:34:25 PDT 2020 x86_64 x86_64 x86_64 GNU/Linux
```

**Ubuntu**
```
grim@reverse:~/pyvmidbg$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
```

**QEMU Config**
```
...
  <qemu:commandline>
    <qemu:arg value="-enable-kvm"/>
    <qemu:arg value="-chardev"/>
    <qemu:arg value="socket,path=/tmp/introspector,id=intro1,reconnect=10"/>
    <qemu:arg value="-object"/>
    <qemu:arg value="introspection,id=kvmi,chardev=intro1"/>
  </qemu:commandline>
...
```

**/etc/libvmi.conf**
```
win10 {
	ostype          = "Windows";
	rekall_profile  = "/home/grim/win10-profile.json";
}
```
win10-profile.json is a rekall profile.

**Comment by Wenzel** [comment]
Hi @waterbear515 

Thanks for this detailed issue, and thanks for trying `pyvmidbg` !

I suspect that the issue lies in `Libvmi`, and not `pyvmidbg`.
What you could try is running a libvmi example that needs to call `vmi_init_complete`, for example `vmi-process-list`:
https://kvm-vmi.github.io/kvm-vmi/master/setup.html#running-libvmi-tools

You could try using a Windows 7 VM, as Windows 10 support in Libvmi highly depends on the build you are running, as there have been plenty of changes in the kernel.

So, check with libvmi first, and if that's confirmed, we can forward this issue to the libvmi repo.

cheers

**Comment by waterbear515** [comment]
Thanks so much for the advice! I just installed the Windows 7 build that can be found in the Vagrant scripts and it worked! It looks like it was a problem with libvmi trying to read from Windows 10 after all.

I also installed the Windows 10 build that's in the Vagrant scripts and it worked as well. Looks like I'll just stick to it and not get a build that's too recent.

Thanks again! Closing because solved.
